File: Python/Types of Matrix/Complex conjugate matrix.py
# Python ≥3.10 is required to run this program
import re
def isNumeric(value):
    try:
        int(value)
        return True
    except:
        return False


def isOperator(operator):
    if operator == '+' or operator == '-':
        return True
    else:
        return False


def conjugate(expression):
    chs = re.split(r'(?=[-+*/])', expression.strip())
    sn = 0  # sum of numerics
    expression = ""
    for i in range(0, len(chs)):
        val = chs[i]
        if len(val) > 0:
            operator = None
            if isOperator(val[0]):
                operator = val[0]
                val = val[1:]
            elif (i == 0):
                operator = '+'

            if len(re.sub(r'^(\d*?[a-zA-Z$])', "", val)) > 0 and isNumeric(val) == False:
                return "error"
            else:
                if isNumeric(val):
                    if int(val) != 0:
                        match operator:
                            case '+': sn += int(val)
                            case '-': sn -= int(val)
                else:
                    match operator:
                        case '+': expression += '-' + val
                        case '-': expression += '+' + val

    if sn == 0 and len(expression) == 0:
        expression = str(sn)
    elif sn != 0:
        expression = str(sn) + expression

    if isOperator(expression[0]) and (expression[0] == '+' or expression == "-0"):
        expression = expression[1:]

    return expression.strip()



A = [
    [ "i", "2+i" ],
    [ "5", "5" ],
    [ "9+7i", "-3-3i" ]
]

print("Matrix: ")
for row in range(0, len(A)):
    for col in range(0, len(A[row])):
        print("\t", A[row][col], "\t", end='')
        if isNumeric(A[row][col]) != True:
            A[row][col] = conjugate(A[row][col])
    print()


# Check if any element/expression has error
error = False
for row in range(0, len(A)):
    for col in range(0, len(A[row])):
        if A[row][col] == "error":
            error = True
            break

print("\nComplex conjugate: ")
if error:
    print("Error! Invalid expression used.")
else:
    for row in range(0, len(A)):
        for col in range(0, len(A[row])):
           print("\t", A[row][col], "\t", end='')
        print()

File: C++/Types of Matrix/Complex conjugate matrix.cpp
#include <iostream>
#include <vector>
#include <regex>
using namespace std;

vector <string> split(string content, regex match){
    vector <string> elements;
    regex_token_iterator<string::iterator>
        iterator(content.begin(), content.end(), match, -1);
    regex_token_iterator<string::iterator> end;
    while (iterator != end)
        elements.push_back(*iterator++);
    return elements;
}
bool isNumeric(string value){
    return value.find_first_not_of("0123456789") == string::npos;
}
bool isOperator(char Operator){
    if(Operator == '+' || Operator == '-'){
        return true;
    }else{
        return false;
    }
}
string conjugate(string expression){
    vector<string> chs = split(expression, regex("(?=[-+*/])"));
    int sn = 0; // sum of numerics
    expression = "";
    for (int i = 0; i < chs.size(); i++) {
        string val = chs[i];
        if(val.length() > 0){
            char Operator = 0;
            if(isOperator(val.at(0))){
                Operator = val.at(0);
                val = val.substr(1);
            } else if (i == 0) {
                Operator = '+';
            }

            if(regex_replace(val, regex("^(\\d*?[a-zA-Z$])"), "").length() > 0 && !isNumeric(val)){
                return "error";
            }else{
                if(isNumeric(val)){
                    if(stoi(val) != 0){
                        switch (Operator) {
                            case '+': sn += stoi(val); break;
                            case '-': sn -= stoi(val); break;
                        }
                    }
                }else{
                    switch (Operator) {
                        case '+': expression += '-' + val; break;
                        case '-': expression += '+' + val; break;
                    }
                }
            }
        }
    }
    if(sn == 0 && expression.length() == 0){
        expression = to_string(sn);
    }else if(sn != 0){
        expression = to_string(sn).append(expression);
    }
    if (isOperator(expression.at(0)) && (expression.at(0) == '+' || expression == "-0")) {
        expression = expression.substr(1);
    }
    return regex_replace(expression, regex("^\\s+|\\s+$"), "");
}

int main(){
    vector <vector<string>> A = {
        { "i", "2+i" },
        { "5", "5" },
        { "9+7i", "-3-3i" }
    };

    cout << "Matrix: " << endl;
    for (int row = 0; row < A.size(); row++) {
        for (int col = 0; col < A[row].size(); col++) {
            cout << "\t" << A[row][col] << "\t";
            if (isNumeric(A[row][col]) != true){
                A[row][col] = conjugate(A[row][col]);
            }
        }
        cout << endl;
    }

    // Check if any element/expression has error
    bool error = false;
    for (int row = 0; row < A.size(); row++) {
        for (int col = 0; col < A[row].size(); col++) {
            if(A[row][col] == "error"){
                error = true;
                break;
            }
        }
    }
    cout <<"\nComplex conjugate: " << endl;
    if(error){
        cout <<"Error! Invalid expression used." << endl;
    } else {
        for (int row = 0; row < A.size(); row++) {
            for (int col = 0; col < A[row].size(); col++) {
                cout <<"\t" << A[row][col] << "\t";
            }
            cout << endl;
        }
    }
}